File: pkg/js/util/model_utils.test.ts
import { getModuleForObjectTypeRelation, isRelationAssignable, isModelModular } from "./model_utils";
import { AuthorizationModel, TypeDefinition, Userset, Usersets, Difference } from "@openfga/sdk";
import { describe } from "@jest/globals";

describe("model_utils", () => {
  describe("getModuleForObjectTypeRelation", () => {
    const typeDefWithModule: TypeDefinition = {
      type: "type1",
      relations: {
        relation1: {},
        relation2: {},
        relation3: {},
        relation4: {},
      },
      metadata: {
        module: "type_module1",
        relations: {
          relation1: { module: "module1" },
          relation2: { module: "" },
          relation3: {},
          relation5: {},
        },
      },
    };

    const typeDefWithoutModule: TypeDefinition = {
      type: "type2",
      relations: {
        relation7: {},
      },
    };

    test("Relation exists and has a module", () => {
      const result = getModuleForObjectTypeRelation(typeDefWithModule, "relation1");
      expect(result).toBe("module1");
    });

    test("Relation exists but has an empty string as a module, type has a module", () => {
      const result = getModuleForObjectTypeRelation(typeDefWithModule, "relation2");
      expect(result).toBe("type_module1");
    });

    test("Relation exists but does not have a module, type has a module", () => {
      const result = getModuleForObjectTypeRelation(typeDefWithModule, "relation3");
      expect(result).toBe("type_module1");
    });

    test("Relation exists but does not have metadata, type has a module", () => {
      const result = getModuleForObjectTypeRelation(typeDefWithModule, "relation4");
      expect(result).toBe("type_module1");
    });

    test("Relation does not exist", () => {
      expect(() => {
        getModuleForObjectTypeRelation(typeDefWithModule, "relation5");
      }).toThrow("relation relation5 does not exist in type type1");
    });

    test("Relation does not exist 2", () => {
      expect(() => {
        getModuleForObjectTypeRelation(typeDefWithModule, "relation6");
      }).toThrow("relation relation6 does not exist in type type1");
    });

    test("Relation exists but does not have a module, type does not have a module", () => {
      const result = getModuleForObjectTypeRelation(typeDefWithoutModule, "relation7");
      expect(result).toBeUndefined();
    });
  });

  describe("isRelationAssignable", () => {
    test("Relation definition has a key 'this'", () => {
      const relDef: Userset = { this: {} };
      expect(isRelationAssignable(relDef)).toBe(true);
    });

    test("Relation definition has a key 'union' with a child that has a key 'this'", () => {
      const relDef: Userset = { union: { child: [{ this: {} }] } as Usersets };
      expect(isRelationAssignable(relDef)).toBe(true);
    });

    test("Relation definition has a key 'intersection' with a child that has a key 'this'", () => {
      const relDef: Userset = { intersection: { child: [{ this: {} }] } as Usersets };
      expect(isRelationAssignable(relDef)).toBe(true);
    });

    test("Relation definition has a key 'difference' with base having a key 'this'", () => {
      const relDef: Userset = { difference: { base: { this: {} }, subtract: {} } as Difference };
      expect(isRelationAssignable(relDef)).toBe(true);
    });

    test("Relation definition has a key 'difference' with subtract having a key 'this'", () => {
      const relDef: Userset = { difference: { base: {}, subtract: { this: {} } } as Difference };
      expect(isRelationAssignable(relDef)).toBe(true);
    });

    test("Relation definition does not have any assignable keys", () => {
      const relDef: Userset = { union: { child: [{ intersection: { child: [{}] } }] } as Usersets };
      expect(isRelationAssignable(relDef)).toBe(false);
    });
  });

  describe("isModelModular", () => {
    test("Model with schema 1.2 and type with module is modular", () => {
      const model = {
        id: "test",
        schema_version: "1.2",
        type_definitions: [
          {
            type: "user",
            relations: {
              viewer: {},
            },
            metadata: {
              module: "user_module",
            },
          },
        ],
      } as AuthorizationModel;
      expect(isModelModular(model)).toBe(true);
    });

    test("Model with schema 1.2 and relation with module is modular", () => {
      const model = {
        id: "test",
        schema_version: "1.2",
        type_definitions: [
          {
            type: "document",
            relations: {
              viewer: {},
            },
            metadata: {
              relations: {
                viewer: { module: "viewer_module" },
              },
            },
          },
        ],
      } as AuthorizationModel;
      expect(isModelModular(model)).toBe(true);
    });

    test("Model with schema 1.1 and type with module is not modular", () => {
      const model = {
        id: "test",
        schema_version: "1.1",
        type_definitions: [
          {
            type: "user",
            relations: {
              viewer: {},
            },
            metadata: {
              module: "user_module",
            },
          },
        ],
      } as AuthorizationModel;
      expect(isModelModular(model)).toBe(false);
    });

    test("Model with schema 1.2 but no modules is not modular", () => {
      const model = {
        id: "test",
        schema_version: "1.2",
        type_definitions: [
          {
            type: "user",
            relations: {
              viewer: {},
            },
          },
        ],
      } as AuthorizationModel;
      expect(isModelModular(model)).toBe(false);
    });

    test("Model with schema 1.2 and empty string module is not modular", () => {
      const model = {
        id: "test",
        schema_version: "1.2",
        type_definitions: [
          {
            type: "user",
            relations: {
              viewer: {},
            },
            metadata: {
              module: "",
            },
          },
        ],
      } as AuthorizationModel;
      expect(isModelModular(model)).toBe(false);
    });

    test("Model with schema 1.2, no type module, but relation with module is modular", () => {
      const model = {
        id: "test",
        schema_version: "1.2",
        type_definitions: [
          {
            type: "document",
            relations: {
              viewer: {},
              editor: {},
            },
            metadata: {
              relations: {
                editor: { module: "editor_module" },
              },
            },
          },
        ],
      } as AuthorizationModel;
      expect(isModelModular(model)).toBe(true);
    });

    test("Model with no schema version throws error", () => {
      const model = {
        id: "test",
        schema_version: undefined as unknown as string,
        type_definitions: [
          {
            type: "user",
            relations: {
              viewer: {},
            },
            metadata: {
              module: "user_module",
            },
          },
        ],
      } as AuthorizationModel;
      expect(() => isModelModular(model)).toThrow("Unsupported schema version: undefined");
    });

    test("Model with unsupported schema version throws error", () => {
      const model = {
        id: "test",
        schema_version: "2.0",
        type_definitions: [
          {
            type: "user",
            relations: {
              viewer: {},
            },
            metadata: {
              module: "user_module",
            },
          },
        ],
      } as AuthorizationModel;
      expect(() => isModelModular(model)).toThrow("Unsupported schema version: 2.0");
    });

    test("Model with empty type definitions is not modular", () => {
      const model = {
        id: "test",
        schema_version: "1.2",
        type_definitions: [],
      } as AuthorizationModel;
      expect(isModelModular(model)).toBe(false);
    });
  });
});

File: pkg/go/utils/model_utils_test.go
package utils

import (
	"testing"

	openfgav1 "github.com/openfga/api/proto/openfga/v1"
)

func TestGetModuleForObjectTypeRelation(t *testing.T) {
	t.Parallel()

	type args struct {
		typeDef  *openfgav1.TypeDefinition
		relation string
	}

	typeDefWithModule := &openfgav1.TypeDefinition{
		Type: "type1",
		Relations: map[string]*openfgav1.Userset{
			"relation1": {},
			"relation2": {},
			"relation3": {},
			"relation4": {},
		},
		Metadata: &openfgav1.Metadata{
			Module: "type_module1",
			Relations: map[string]*openfgav1.RelationMetadata{
				"relation1": {Module: "module1"},
				"relation2": {Module: ""},
				"relation3": {},
				"relation5": {},
			},
		},
	}

	typeDefWithoutModule := &openfgav1.TypeDefinition{
		Type: "type2",
		Relations: map[string]*openfgav1.Userset{
			"relation7": {},
		},
	}

	tests := []struct {
		name    string
		args    args
		want    string
		wantErr bool
	}{
		{
			name: "Relation exists and has a module",
			args: args{
				typeDef:  typeDefWithModule,
				relation: "relation1",
			},
			want:    "module1",
			wantErr: false,
		},
		{
			name: "Relation exists but has an empty string as a module, type has a module",
			args: args{
				typeDef:  typeDefWithModule,
				relation: "relation2",
			},
			want:    "type_module1",
			wantErr: false,
		},
		{
			name: "Relation exists but does not have a module, type has a module",
			args: args{
				typeDef:  typeDefWithModule,
				relation: "relation3",
			},
			want:    "type_module1",
			wantErr: false,
		},
		{
			name: "Relation exists but does not have metadata, type has a module",
			args: args{
				typeDef:  typeDefWithModule,
				relation: "relation4",
			},
			want:    "type_module1",
			wantErr: false,
		},
		{
			name: "Relation does not exist",
			args: args{
				typeDef:  typeDefWithModule,
				relation: "relation5",
			},
			want:    "",
			wantErr: true,
		},
		{
			name: "Relation does not exist 2",
			args: args{
				typeDef:  typeDefWithModule,
				relation: "relation6",
			},
			want:    "",
			wantErr: true,
		},
		{
			name: "Relation exists but does not have a module, type does not have a module",
			args: args{
				typeDef:  typeDefWithoutModule,
				relation: "relation7",
			},
			want:    "",
			wantErr: false,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			got, err := GetModuleForObjectTypeRelation(tt.args.typeDef, tt.args.relation)
			if (err != nil) != tt.wantErr {
				t.Errorf("GetModuleForObjectTypeRelation() error = %v, wantErr %v", err, tt.wantErr)

				return
			}

			if got != tt.want {
				t.Errorf("GetModuleForObjectTypeRelation() got = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestIsRelationAssignable(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name     string
		relDef   *openfgav1.Userset
		expected bool
	}{
		{
			name:     "Relation definition has a key 'this'",
			relDef:   &openfgav1.Userset{Userset: &openfgav1.Userset_This{}},
			expected: true,
		},
		{
			name: "Relation definition has a key 'union' with a child that has a key 'this'",
			relDef: &openfgav1.Userset{Userset: &openfgav1.Userset_Union{
				Union: &openfgav1.Usersets{
					Child: []*openfgav1.Userset{
						{Userset: &openfgav1.Userset_This{}},
					},
				},
			}},
			expected: true,
		},
		{
			name: "Relation definition has a key 'intersection' with a child that has a key 'this'",
			relDef: &openfgav1.Userset{Userset: &openfgav1.Userset_Intersection{
				Intersection: &openfgav1.Usersets{
					Child: []*openfgav1.Userset{
						{Userset: &openfgav1.Userset_This{}},
					},
				},
			}},
			expected: true,
		},
		{
			name: "Relation definition has a key 'difference' with base having a key 'this'",
			relDef: &openfgav1.Userset{Userset: &openfgav1.Userset_Difference{
				Difference: &openfgav1.Difference{
					Base:     &openfgav1.Userset{Userset: &openfgav1.Userset_This{}},
					Subtract: &openfgav1.Userset{},
				},
			}},
			expected: true,
		},
		{
			name: "Relation definition has a key 'difference' with subtract having a key 'this'",
			relDef: &openfgav1.Userset{Userset: &openfgav1.Userset_Difference{
				Difference: &openfgav1.Difference{
					Base:     &openfgav1.Userset{},
					Subtract: &openfgav1.Userset{Userset: &openfgav1.Userset_This{}},
				},
			}},
			expected: true,
		},
		{
			name: "Relation definition does not have any assignable keys",
			relDef: &openfgav1.Userset{Userset: &openfgav1.Userset_Union{
				Union: &openfgav1.Usersets{
					Child: []*openfgav1.Userset{
						{Userset: &openfgav1.Userset_Intersection{
							Intersection: &openfgav1.Usersets{
								Child: []*openfgav1.Userset{
									{},
								},
							},
						}},
					},
				},
			}},
			expected: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			result := IsRelationAssignable(tt.relDef)
			if result != tt.expected {
				t.Errorf("expected %v, got %v", tt.expected, result)
			}
		})
	}
}
